gerritlab/main.py

```diff
@@ -126,91 +126,50 @@ def cancel_prev_pipelines(commit):
+            mr.source_branch: mr for mr in merge_request.get_all_merge_requests(
+                remote, local_branch)
+        }
 
-    # Update the existing MRs from the end of the MR dependency chain.
-    updated_commits = []
-    updated_mrs = []
-    for mr in reversed(current_mr_chain):
-        for c in commits_data:
-            if c.source_branch == mr.source_branch:
-                # Delay the update if the updated target branch doesn't exist,
-                # meaning a new MR will be created off of that target branch.
-                updated_target_branch_exists = "{}/{}".format(
-                    remote.name,
-                    c.target_branch) in [ref.name for ref in repo.references]
-                if not updated_target_branch_exists:
-                    continue
-                # Do nothing if the commit doesn't contain any new changes.
-                if can_skip_ci(c.commit, c.source_branch):
-                    continue
-                # Update the target branch of this MR.
-                title, desp = utils.get_msg_title_description(c.commit.message)
-                mr.update(
-                    source_branch=c.source_branch,
-                    target_branch=c.target_branch, title=title,
-                    description=desp)
-                cancel_prev_pipelines(c.commit)
-                # Update the remote branch.
-                remote.push(
-                    refspec="{}:refs/heads/{}".format(
-                        c.commit.hexsha, c.source_branch), force=True)
-                updated_commits.append(c)
-                updated_mrs.append(mr)
-                break
-
-    # Create new MRs.
-    commits_data = [c for c in commits_data if c not in updated_commits]
     new_mrs = []
+    updated_mrs = []
+    
     for c in commits_data:
         title, desp = utils.get_msg_title_description(c.commit.message)
-        mr = merge_request.get_merge_request(remote, c.source_branch)
-        if mr is not None:
-            # Do nothing if the commit doesn't contain any new changes.
-            if can_skip_ci(c.commit, c.source_branch):
-                continue
-            cancel_prev_pipelines(c.commit)
-            # Update the remote branch.
-            remote.push(
-                refspec="{}:refs/heads/{}".format(
-                    c.commit.hexsha, c.source_branch), force=True)
-            # Update the MR if it already exits.
-            mr.update(
-                source_branch=c.source_branch,
-                target_branch=c.target_branch, title=title,
-                description=desp)
-            updated_mrs.append(mr)
+        
+        mr = current_mrs_by_source_branch.get(c.source_branch)
+        if mr:
+            # Update the MR if needed
+            if (mr.source_branch != c.source_branch or 
+                mr.target_branch != c.target_branch or 
+                mr._title != title or 
+                mr._description != desp.strip()):
```

Comment:
I added something like this.  Btw `if mr and merge_request.is_outdated()` cannot be a single `if` because if the expression is false, the `else` will attempt to create an MR (which is the not desired in the case where the MR exists but does not need to be updated).